ddtrace/contrib/internal/langchain/patch.py

```diff
@@ -965,6 +965,170 @@ def traced_similarity_search(langchain, pin, func, instance, args, kwargs):
     return documents
 
 
+@with_traced_module
+def traced_base_tool_invoke(langchain, pin, func, instance, args, kwargs):
+    integration = langchain._datadog_integration
+    tool_input = get_argument_value(args, kwargs, 0, "input")
+    config = get_argument_value(args, kwargs, 1, "config") if len(args) >= 2 else None
+
+    span = integration.trace(
+        pin,
+        "%s.%s.%s" % (func.__module__, func.__class__.__name__, func.__self__.__class__.__name__),
+        interface_type="tool",
+        submit_to_llmobs=True,
+    )
+
+    tool_output = None
+    try:
+        if instance.name:
+            span.set_tag_str("langchain.request.tool.name", integration.trunc(str(instance.name)))
+        if integration.is_pc_sampled_span(span):
+            if instance.description:
+                span.set_tag_str("langchain.request.tool.description", integration.trunc(str(instance.description)))
+            if instance.return_direct:
+                span.set_tag_str("langchain.request.tool.return_direct", integration.trunc(str(instance.return_direct)))
+            if instance.verbose:
+                span.set_tag_str("langchain.request.tool.verbose", integration.trunc(str(instance.verbose)))
+            if instance.metadata:
+                for key, value in instance.metadata.items():
+                    span.set_tag_str("langchain.request.tool.metadata.%s" % key, integration.trunc(str(value)))
+            if instance.tags:
+                for idx, tag in enumerate(instance.tags):
+                    span.set_tag_str("langchain.request.tool.tags.%d" % idx, integration.trunc(str(tag)))
+            if instance.handle_tool_error:
+                span.set_tag_str(
+                    "langchain.request.tool.handle_tool_error",
+                    integration.trunc(str(instance.handle_tool_error)),
+                )
+            if instance.handle_validation_error:
+                span.set_tag_str(
+                    "langchain.request.tool.handle_validation_error",
+                    integration.trunc(str(instance.handle_validation_error)),
+                )
+            if instance.response_format:
+                span.set_tag_str(
+                    "langchain.request.tool.response_format", integration.trunc(str(instance.response_format))
+                )
+            if tool_input:
+                span.set_tag_str("langchain.request.input", integration.trunc(str(tool_input)))
+            if config:
```

Comment:
Is this enough or should I parse the configs one by one and add them

Reply:
1. We don't need to truncate config as that isn't a direct I/O to the tool
2. We can probably just json dump the config as it is a `TypedDict` ([see source](https://python.langchain.com/v0.2/api_reference/_modules/langchain_core/runnables/config.html#RunnableConfig))